subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepository.java

```diff
@@ -112,78 +133,125 @@ public String toString() {
-        public void resolveArtifact(ComponentArtifactMetadata artifact, ModuleSource moduleSource, BuildableArtifactResolveResult result) {
+        private <E extends Throwable, R extends ErroringResolveResult<E>> void performOperationWithRetries(R result, Runnable operation, Transformer<E, Throwable> onError) {
             if (repositoryBlacklister.isBlacklisted(repositoryId)) {
-                result.failed(new ArtifactResolveException(artifact.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE));
+                result.failed(onError.transform(null));
                 return;
             }
 
-            try {
-                delegate.resolveArtifact(artifact, moduleSource, result);
-            } catch (Throwable throwable) {
-                repositoryBlacklister.blacklistRepository(repositoryId, throwable);
-                result.failed(new ArtifactResolveException(artifact.getId(), throwable));
+            tryResolveAndMaybeBlacklist(result, operation, onError);
+        }
+
+        private <E extends Throwable, R extends ErroringResolveResult<E>> void tryResolveAndMaybeBlacklist(R result, Runnable operation, Transformer<E, Throwable> onError) {
+            tryResolveAndMaybeBlacklist(result, () -> {
+                operation.run();
+                return null;
+            }, onError);
+        }
+
+        private <E extends Throwable, R extends ErroringResolveResult<E>> void tryResolveAndMaybeBlacklist(R result, Callable<E> operation, Transformer<E, Throwable> onError) {
+            int retries = 0;
+            int backoff = initialBackOff;
+
+            while (retries < maxRetriesCount) {
+                retries++;
+                E failure;
+                Throwable unexpectedFailure = null;
+                try {
+                    failure = operation.call();
+                    if (failure == null) {
+                        if (retries > 1) {
+                            LOGGER.debug("Successfully fetched external resource after {} retries", retries - 1);
+                        }
+                        return;
+                    }
+                } catch (Throwable throwable) {
+                    unexpectedFailure = throwable;
+                    failure = onError.transform(throwable);
+                }
+                if (retries == maxRetriesCount) {
+                    if (unexpectedFailure != null) {
+                        repositoryBlacklister.blacklistRepository(repositoryId, unexpectedFailure);
+                    }
+                    result.failed(failure);
+                    break;
+                } else {
+                    LOGGER.debug("Error while accessing remote repository {}. Waiting {}ms before next retry. {} retries left", repositoryId, backoff, maxRetriesCount - retries);
+                    try {
+                        Thread.sleep(backoff);
+                        backoff *= 2;
+                    } catch (InterruptedException e) {
+                        break;
```

Comment:
❓ Shouldn't you be restoring the interrupted status here if not propagating the exception?

Reply:
Changed to rethrow instead.